examples/py_triple.rs

```diff
@@ -0,0 +1,411 @@
+
+        let perm = {
+            // Convert advice columns into an "any" columns.
+            let cols: [Column<Any>; 3] = [l_col.into(), r_col.into(), o_col.into()];
+            Permutation::new(cs, &cols)
+        };
+
+        let s_add = cs.selector();
+        let s_mul = cs.selector();
+        let s_pub = cs.selector();
+
+        // Define the addition gate.
+        //
+        // | l_col | r_col | o_col | s_add |
+        // |-------|-------|-------|-------|
+        // |   l   |   r   |   o   | s_add |
+        //
+        // Constraint: s_add*l + s_add*r = s_add*o
+        cs.create_gate("add", |cs| {
+            let l = cs.query_advice(l_col, Rotation::cur());
+            let r = cs.query_advice(r_col, Rotation::cur());
+            let o = cs.query_advice(o_col, Rotation::cur());
+            let s_add = cs.query_selector(s_add, Rotation::cur());
+            s_add * (l + r - o)
+        });
+
+        // Define the multiplication gate.
+        //
+        // | l_col | r_col | o_col | s_mul |
+        // |-------|-------|-------|-------|
+        // |   l   |   r   |   o   | s_mul |
+        //
+        // Constraint: s_mul*l*r = s_mul*o
+        cs.create_gate("mul", |cs| {
+            let l = cs.query_advice(l_col, Rotation::cur());
+            let r = cs.query_advice(r_col, Rotation::cur());
+            let o = cs.query_advice(o_col, Rotation::cur());
+            let s_mul = cs.query_selector(s_mul, Rotation::cur());
+            s_mul * (l * r - o)
+        });
+
+        // Define the public input gate.
+        //
+        // | l_col | pub_col | s_pub |
+        // |-------|---------|-------|
+        // |   l   |    pi   | s_pub |
+        //
+        // Constraint: s_pub*l = s_pub*pi
+        cs.create_gate("public input", |cs| {
+            let l = cs.query_advice(l_col, Rotation::cur());
+            let pi = cs.query_instance(pub_col, Rotation::cur());
+            let s_pub = cs.query_selector(s_pub, Rotation::cur());
+            s_pub * (l - pi)
+        });
+
+        MyChipConfig {
+            l_col,
+            r_col,
+            o_col,
+            perm,
+            s_add,
+            s_mul,
+            s_pub,
+        }
+    }
+
+    // In the next available row, writes `a` into the row's left cell and `b` into the row's right
+    // cell.
+    fn alloc_private_inputs(
+        &self,
+        mut layouter: impl Layouter<Fp>,
+        a: Fp,
+        b: Fp,
+    ) -> Result<(Alloc, Alloc), Error> {
+        let config = self.config();
+        let mut a_alloc: Option<Alloc> = None;
+        let mut b_alloc: Option<Alloc> = None;
+
+        layouter.assign_region(
+            || "load private inputs",
+            |mut region| {
+                let row_offset = 0;
+                let a_cell = region.assign_advice(
+                    || "private input 'a'",
+                    config.l_col,
+                    row_offset,
+                    || Ok(a),
```

Comment:
We need to ensure that variables are only unwrapped inside these value assignment closures. The closures are only called when we are guaranteed to have a value (e.g. this advice closure will only be called during proving, not keygen).
```suggestion
                    || a.ok_or(Error::SynthesisError),
```